### PR TITLE
chore(profile::jenkinscontroller) Cleanup plugins and JCasC management

### PR DESCRIPTION
**Pull request by dduportal**
This PR fixes a few elements in the `jenkinscontroller` profile which will is not expecting to impact our 3 controllers (ci.jio, cert.ci and trusted.ci) BUT could accidentally.

The fixed elements are related to the provisioning of Jenkins plugin installations and JCasC configuration files.
The goal is to fix the following error or edge cases, anticipating for the support of EC2 Fleet plugin and also migration to Ansible:

1. (local dev/tests) Vagrant local VM does not provision and need 2-3 reprovisioning until it reaches a stable state. 
2. (production/local dev/tests) when a JCasC configuration file require a given plugin, the file is often written before the plugin is installed, which may lead to endless loop of restarts (case with ec2 fleet but also have been the case with artifact manager, azure VM and EC2 in the past) requiring a manual operation.

5 changes are expected to solve these 2 problems, they are all well-defined within distinct commits:

- (chore) Avoid specifying a `KUBECONFIG` empty variable when not using credential-less Kubernetes agents. Avoid bad null-iness checks on downstream setups.
- (chore) install plugins with distinct containers than the Jenkins controller. Allows to break endless loops when failing. The additional overhead is not worth mentioning.
  - Note: not removing (`docker run --rm ...`) containers as we have a GC. Useful to keep them for logs when it fails to install a given plugin.
- (chore) improve plugins auto-detection from JCasC configuration (a new plugin is covered, fixed the nested map merge in hieradata and use the resolved jcasc configuration which is more efficient)
- (chore) add a dummy SSH key credential to break endless loop of restarts
- (nit) Removed a now unused resource

⚠️ Do not merge once reviewed: I want to carefully deploy this manually on each controller to avoid weird behavior ⚠️ 

----

Notes: 
- A last commit is added for unit tests (I wrote them at the beginning but forgot to commit it).
- These changes are required ofr EC2 fleet support (see incoming PR definition) to support setting up AWS credential on vagrant

### CODE REVIEW

File: dist/profile/manifests/jenkinscontroller.pp [hunk 1]
```diff
@@ -434,13 +430,21 @@
...
-    # Only install plugins after we've secured Jenkins, that seems reasonable
-    require => [
-      File[$groovy_d],
-    ],
+  $all_plugins.each |$plugin| {
+    exec { "install-plugin-${plugin}":
+      ## Check for plugin presence on the HOST (e.g. with the jenkins home in "/var/lib/jenkins" on the filesystem)
+      unless    => "/usr/bin/test -f /var/lib/jenkins/plugins/${plugin}.jpi || /usr/bin/test -f /var/lib/jenkins/plugins/${plugin}.hpi",
+      ## Install the plugin (if needed) in the container, e.g. with the jenkins home mounted in /var/jenkins_home
+      # command => "docker exec jenkins jenkins-plugin-cli --plugins ${name} --plugin-download-directory /var/jenkins_home/plugins",
```
Comment (lemeurherve) [inline dist/profile/manifests/jenkinscontroller.pp]:
```suggestion
```
nit: remove commented command

File: dist/profile/manifests/jenkinscontroller.pp [hunk 2]
```diff
@@ -434,13 +430,21 @@
...
+      command   => "docker run -t --entrypoint=jenkins-plugin-cli --env=CACHE_DIR=/tmp --restart=no --volume=${jenkins_home}:/var/jenkins_home:rw --user=$(id -u jenkins):$(id -g jenkins) ${docker_image}:${docker_tag}  --plugins ${plugin} --plugin-download-directory /var/jenkins_home/plugins",
```
Comment (lemeurherve) [inline dist/profile/manifests/jenkinscontroller.pp]:
```suggestion
      command   => "docker run -t --entrypoint=jenkins-plugin-cli --env=CACHE_DIR=/tmp --restart=no --volume=${jenkins_home}:/var/jenkins_home:rw --user=$(id -u jenkins):$(id -g jenkins) ${docker_image}:${docker_tag} --plugins ${plugin} --plugin-download-directory /var/jenkins_home/plugins",
```
nit: extraneous space removal

**Review by lemeurherve — DISMISSED** [review]
LGTM, thanks!

File: spec/classes/profile/jenkinscontroller_spec.rb [hunk 3]
```diff
@@ -129,7 +139,12 @@
     it "should mount the kubeconfigs directory in the container with an updated KUBECONFIG" do
       expect(subject).to contain_docker__run("jenkins").with({
         :pull_on_start => true,
-        :volumes => ["/var/lib/jenkins:/var/jenkins_home:rw", nil, "/var/jenkins_kubeconfigs:/var/jenkins_kubeconfigs:ro"],
+        :volumes => [
+          "/var/lib/jenkins:/var/jenkins_home:rw",
+          nil,
+          "/var/jenkins_kubeconfigs:/var/jenkins_kubeconfigs:ro",
+          "/var/run/jenkins-secrets:/run/secrets:ro",
```
Comment (dduportal) [inline spec/classes/profile/jenkinscontroller_spec.rb:146]:
```suggestion
```

This change was not expected in this PR. Incoming hotfix.

File: dist/profile/manifests/jenkinscontroller.pp [hunk 4]
```diff
@@ -386,42 +383,41 @@
...
-    ],
+      $kubeconfig_envvar_value,
+    ].map |$item| { if $item != '' { $item } },
     ports            => ['8080:8080', '50000:50000'],
-    volumes          => concat(["${jenkins_home}:/var/jenkins_home:rw"],$awscli_container_volume,$kubeconfig_container_volume).map |$item| { if $item != '' { $item } },
+    volumes          => concat([
+      "${jenkins_home}:/var/jenkins_home:rw"],
+      $awscli_container_volume,
+      $kubeconfig_container_volume,
+      '/var/run/jenkins-secrets:/run/secrets:ro',
```
Comment (dduportal) [inline dist/profile/manifests/jenkinscontroller.pp:393]:
```suggestion
```

This change was not expected in this PR. Incoming hotfix.

File: spec/classes/profile/jenkinscontroller_spec.rb [hunk 5]
```diff
@@ -35,7 +35,12 @@
       it "should define a suitable docker::run" do
         expect(subject).to contain_docker__run("jenkins").with({
           :pull_on_start => true,
-          :volumes => ["/var/lib/jenkins:/var/jenkins_home:rw", nil, nil],
+          :volumes => [
+            "/var/lib/jenkins:/var/jenkins_home:rw",
+            nil,
+            nil,
+            "/var/run/jenkins-secrets:/run/secrets:ro",
```
Comment (dduportal) [inline spec/classes/profile/jenkinscontroller_spec.rb:42]:
```suggestion
```

This change was not expected in this PR. Incoming hotfix.

File: spec/classes/profile/jenkinscontroller_spec.rb [hunk 6]
```diff
@@ -93,15 +98,20 @@
     it "should mount the aws CLI installation directory in the container with an updated PATH" do
       expect(subject).to contain_docker__run("jenkins").with({
         :pull_on_start => true,
-        :volumes => ["/var/lib/jenkins:/var/jenkins_home:rw", "/var/awscli:/var/awscli:ro", nil],
+        :volumes => [
+          "/var/lib/jenkins:/var/jenkins_home:rw",
+          "/var/awscli:/var/awscli:ro",
+          nil,
+          "/var/run/jenkins-secrets:/run/secrets:ro",
```
Comment (dduportal) [inline spec/classes/profile/jenkinscontroller_spec.rb:105]:
```suggestion
```

This change was not expected in this PR. Incoming hotfix.

File: dist/profile/manifests/jenkinscontroller.pp (hunk 4 above)
Comment (dduportal) [reply dist/profile/manifests/jenkinscontroller.pp:393]: https://github.com/jenkins-infra/jenkins-infra/commit/2320e363a0e2c1f716796e43f4e42b0b386e1d60

File: spec/classes/profile/jenkinscontroller_spec.rb (hunk 5 above)
Comment (dduportal) [reply spec/classes/profile/jenkinscontroller_spec.rb:42]: https://github.com/jenkins-infra/jenkins-infra/commit/2320e363a0e2c1f716796e43f4e42b0b386e1d60

File: spec/classes/profile/jenkinscontroller_spec.rb (hunk 6 above)
Comment (dduportal) [reply spec/classes/profile/jenkinscontroller_spec.rb:105]: https://github.com/jenkins-infra/jenkins-infra/commit/2320e363a0e2c1f716796e43f4e42b0b386e1d60

File: spec/classes/profile/jenkinscontroller_spec.rb (hunk 3 above)
Comment (dduportal) [reply spec/classes/profile/jenkinscontroller_spec.rb:146]: https://github.com/jenkins-infra/jenkins-infra/commit/2320e363a0e2c1f716796e43f4e42b0b386e1d60

**Comment by dduportal** [comment]
- Required a hotfix: https://github.com/jenkins-infra/jenkins-infra/commit/2320e363a0e2c1f716796e43f4e42b0b386e1d60 (unwanted change in this PR)
- With the hotfix, `noop` results shows expected diffs:
  - No changes planned on ci.jenkins.io
  - removal of the `KUBECONFIG` empty environment variable on cert.ci and trusted.ci
